feat(minifier): fold simple literals passed to unary `+`

sapphi-red opened the pull request:
Added support for folding `+[]` and `+/any-regex/`.

`+` uses `ToNumber` and `ToNumber` uses `ToPrimitive` for objects. `ToPrimitive` for `[]` returns `""`, for `/any-regex/` it returns `"/any-regex/"`.

**References**
- [Spec of unary `+`](https://tc39.es/ecma262/multipage/ecmascript-language-expressions.html#sec-unary-plus-operator-runtime-semantics-evaluation)
- [Spec of `ToPrimitive`](https://tc39.es/ecma262/multipage/abstract-operations.html#sec-toprimitive)

sapphi-red commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8943?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#8944** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8944?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#8943** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8943?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8943?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#8942** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8942?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#8941** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8941?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=8943).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

sapphi-red commented on crates/oxc_ecmascript/src/constant_evaluation/mod.rs:
```diff
@@ -145,7 +145,20 @@ pub trait ConstantEvaluation<'a>: MayHaveSideEffects {
+                } else {
+                    None
+                }
+            }
```
I wonder if it makes sense to move these to
https://github.com/oxc-project/oxc/blob/515eb525027780e7449db315fc5c7c435e18b21f/crates/oxc_ecmascript/src/to_number.rs#L11

Boshen replied:
Probably? `oxc_ecmascript` implements the spec.

sapphi-red replied:
OK. I'll move them later on.

sapphi-red committed:
feat(minifier): fold simple literals passed to unary `+` (#8943)

Added support for folding `+[]` and `+/any-regex/`.

`+` uses `ToNumber` and `ToNumber` uses `ToPrimitive` for objects. `ToPrimitive` for `[]` returns `""`, for `/any-regex/` it returns `"/any-regex/"`.

**References**
- [Spec of unary `+`](https://tc39.es/ecma262/multipage/ecmascript-language-expressions.html#sec-unary-plus-operator-runtime-semantics-evaluation)
- [Spec of `ToPrimitive`](https://tc39.es/ecma262/multipage/abstract-operations.html#sec-toprimitive)
